### PR TITLE
Fix: NotebookProgressCallback crash when evaluating with the Trainer

### PR DESCRIPTION
**Pull request by Charly21r**
# What does this PR do?

Fixes #44936 

This PR fixes an issue with `NotebookProgressCallback` in the `Trainer` where calling evaluate() before or after training would crash due to the training tracker being `None`. The callback now properly handles evaluation even if training has not yet started or if it has already finished, ensuring metrics can be computed and displayed.

Previously, the `on_evaluate` method assumed that `self.training_tracker` was always initialized, but:
- Before training: `self.training_tracker` has not being initialised by `on_train_begin` yet.
- After training: `on_train_end` sets `self.training_tracker` to `None`, so calling `on_evaluate` afterwards would fail.

**Fix:** `on_evaluate` now checks whether self.training_tracker exists before using it, and safely handles cases where it is `None`. This prevents crashes and ensures evaluation can run regardless of training state.

Additionally, new unit tests were added to ensure that evaluation works in this scenario, and existing notebook callback tests were updated to cover this case. This improves robustness of notebook-based workflows, especially in Jupyter or Colab environments.

## Code Agent Policy

- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Did you write any new necessary tests?

## Who can review?

@SunMarc 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by SunMarc** [comment]
does this fix your issue @HenrikEilers ?

**Comment by Charly21r** [comment]
@SunMarc Just checking in, happy to make any changes if needed. If the original reporter isn’t available, I can also provide more details or tests to help validate the fix.

**Comment by HenrikEilers** [comment]
> does this fix your issue @HenrikEilers ?

I cant really test it right now but if it now allows for trainer.evaluate() to be called after training then yes

**Comment by HenrikEilers** [comment]
Is there anything left to do for us to get the pr aproved?

**Comment by Charly21r** [comment]
Just checking in again, this PR should resolve the original issue by allowing trainer.evaluate() after training, as discussed above. Tests are passing on my side, and I’m happy to add more coverage if needed.

@SunMarc I'd really appreciate a review when you have time :)

File: src/transformers/utils/notebook.py [hunk 1]
```diff
@@ -351,7 +351,10 @@ def on_log(self, args, state, control, logs=None, **kwargs):
             tt.write_line(values)
 
     def on_evaluate(self, args, state, control, metrics=None, **kwargs):
-        tt = _require(self.training_tracker, "on_train_begin must be called before on_evaluate")
+        if self.training_tracker is None:
+            return control
```
Comment (SunMarc) [inline src/transformers/utils/notebook.py]: Ok this will work but we are not outputing anything in this case no like `tt.write_line(values)`. Can you check what is the output that we get and if it makes sense, maybe we should add it.

**Review by SunMarc — COMMENTED** [review]
Thanks ! Left a comment

File: src/transformers/utils/notebook.py (hunk 1 above)
Comment (Charly21r) [reply src/transformers/utils/notebook.py]: Good point, I’ve updated `on_evaluate` to display the metrics as a standalone HTML table when there's no training tracker, so the user still sees the output. I also included the `first_column` computation into `on_evaluate` directly so it doesn't depend on `on_train_begin having run.

Comment (SunMarc) [reply src/transformers/utils/notebook.py]: can you show me what you get ?

File: tests/trainer/test_trainer_callback.py [hunk 2]
```diff
@@ -1269,3 +1273,75 @@ def state(self):
 
         self.assertEqual(instance.name, "test")
         self.assertEqual(instance.counter, 5)
+
+
+@require_torch
+@unittest.skipUnless(IPYTHON_AVAILABLE, "IPython is required for NotebookProgressCallback")
```
Comment (SunMarc) [inline tests/trainer/test_trainer_callback.py]: let's create a `is_ipython_available` + `require_ipython` as we do for many other lib instead of this

File: src/transformers/utils/notebook.py [hunk 3]
```diff
@@ -351,7 +351,7 @@ def on_log(self, args, state, control, logs=None, **kwargs):
             tt.write_line(values)
 
     def on_evaluate(self, args, state, control, metrics=None, **kwargs):
-        tt = _require(self.training_tracker, "on_train_begin must be called before on_evaluate")
+        self.first_column = "Epoch" if args.eval_strategy == IntervalStrategy.EPOCH else "Step"
```
Comment (SunMarc) [inline src/transformers/utils/notebook.py:356]: why we need to overwrite that ? we shouldn't have to

**Review by SunMarc — COMMENTED** [review]
Thanks, a few nits

File: src/transformers/utils/notebook.py (hunk 1 above)
Comment (Charly21r) [reply src/transformers/utils/notebook.py]: **Calling `evaluate()` after `train()`:**
Screenshot 2026-04-09 at 19 40 01

**Calling `evaluate()` before `train()`:**
Screenshot 2026-04-09 at 19 44 35


I noticed `Model Preparation Time` shows up as a column when calling it before train, should I filter it out along with the other runtime metrics?

File: src/transformers/utils/notebook.py (hunk 3 above)
Comment (Charly21r) [reply src/transformers/utils/notebook.py:356]: Because `on_evaluate` can also be called before `on_train_begin` (which is the bug this PR fixes), but `self.first_column` wouldn't exist yet since it's only initialized in `on_train_begin`. Another option would be to move the initialization to `__init__` with a default of "Step" instead, so `on_evaluate` doesn't need to overwrite it. Defaulting to "Step" could make sense here since if training hasn't started, there are no epochs to reference. So I can do that if you prefer it.

File: tests/trainer/test_trainer_callback.py (hunk 2 above)
Comment (Charly21r) [reply tests/trainer/test_trainer_callback.py]: Done!

File: src/transformers/utils/notebook.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/utils/notebook.py]: Yeah, Model preparation time shouldn't show up. We shouldn't necessarily filter this one in particular but find out why it shows up when it doesn't when calling evaluate after train. Thanks for testing !

File: src/transformers/utils/notebook.py (hunk 3 above)
Comment (SunMarc) [reply src/transformers/utils/notebook.py:356]: understood. let's keep what you did, maybe just add a comment about that above

**Review by SunMarc — APPROVED** [review]
Thanks ! Just fix the issue about the logs that shows up incorrectly when doing evaluation first and we can merge this

File: src/transformers/utils/notebook.py (hunk 1 above)
Comment (Charly21r) [reply src/transformers/utils/notebook.py]: Found it. The trainer adds `eval_model_preparation_time` to the metrics dict when the model hasn't been prepared yet (`self.accelerator._models` is empty), which only happens when `evaluate()` is called before `train()`. After training, the model is already prepared so the metric is never added. The fix is just adding `metrics.pop(f\"{metric_key_prefix}_model_preparation_time\", None)` alongside the other filtered metrics.

File: src/transformers/utils/notebook.py (hunk 3 above)
Comment (Charly21r) [reply src/transformers/utils/notebook.py:356]: Done!

File: src/transformers/utils/notebook.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/utils/notebook.py]: SG !

**Review by SunMarc — APPROVED** [review]
Thanks for iterating !
